### PR TITLE
add rng option to jittablemodule

### PR DESCRIPTION
**Pull request by qihqi**
Add an option for JittableModule to take in rng. This Pr addresses the issue in https://github.com/google/torchax/issues/17 where models with dropout want to use different rng behavior every run.

### CODE REVIEW

**Review by QiliangCui — APPROVED** [review]

**Review by wdhongtw — CHANGES_REQUESTED** [review]
Need some sort of change. See other comments.

**Comment by qihqi** [comment]
Hi @wdhongtw 

The changes you propose makes sense. I made the changes accordingly. So now forward doesn't chagne signature but gets the rng from env and pass it down to functional_call; and functional_call is being jitted so it will not hardcde a rng.

**Review by wdhongtw — COMMENTED** [review]
Still some questions. 🙂‍↕️

File: torchax/interop.py [hunk 1]
```diff
@@ -103,11 +111,9 @@ def __class__(self):
     # isinstance(jittable_module, self._model.__class__) works
     return self._model.__class__
 
-  def __call__(self, *args, **kwargs):
-    return self.forward(*args, **kwargs)
```
Comment (wdhongtw) [inline torchax/interop.py:111]: Not quite sure about this removal of `__call__`, I think that resolve some issue around backward pass?

If so, I thinks this removal deserves a separated PR. 👍

File: torchax/interop.py [hunk 2]
```diff
@@ -124,10 +130,19 @@ def functional_call(self, method_or_name, params, buffers, *args, **kwargs):
         )
       method = method_or_name
       args = (self._model,) + args
-    with torch_stateless._reparametrize_module(self._model, params_copy):
-      res = method(*args, **kwargs)
+    with self._env as env:
```
Comment (wdhongtw) [inline torchax/interop.py:136]: Is entering the Environment required here? Shouldn't we enforce users to activate the context explicitly through `torchax.enable_globally` or `with torchax.default_env` as the usage guide in README?

File: torchax/interop.py [hunk 3]
```diff
@@ -124,10 +130,19 @@ def functional_call(self, method_or_name, params, buffers, *args, **kwargs):
...
+      with (
+        env.override_property(prng=rng),
+        torch_stateless._reparametrize_module(self._model, params_copy),
+      ):
+        res = method(*args, **kwargs)
     return res
 
+  @contextlib.contextmanager
+  def with_rng(self, rng):
+    with self._env.override_property(prng=_jax_view(rng)):
```
Comment (wdhongtw) [inline torchax/interop.py]: What's the expected type of rng here? Can we add explicit type hint for this rng?

If it's a torch.Tensor, then the usage in test file seems incorrect. If it's a jax.Array, then the `_jax_view` is redundent.

File: torchax/interop.py (hunk 2 above)
Comment (qihqi) [reply torchax/interop.py:136]: Because the class constructor takes environment as input. So we need to use the env that was originally passed in. The user can pass in default_env in the constructor.

So I want to start encouraging the pattern of passing env around -- potentially distinct envs, explicitly. So say 2 JittableModules dont need to run serially just because they both use rng.

File: torchax/interop.py (hunk 1 above)
Comment (qihqi) [reply torchax/interop.py:111]: done

File: torchax/interop.py (hunk 3 above)
Comment (qihqi) [reply torchax/interop.py]: done.

File: torchax/interop.py (hunk 2 above)
Comment (wdhongtw) [reply torchax/interop.py:136]: Thanks Qi Han, I think this is going to be merged soon.

---

I agree the idea of passing env into `JittableModule`, the default value None is for backward compatibility.
(Also agree with the potentially distinct envs, but we still need to be careful with the thread local state hidden inside PyTorch framework. I think you have more expertise in this part. 👍)

I'm just wondering the responsibility for activating the environment (and register the dispatch handler).
If we still rely on user to activate the environment, the computation works and that conform the current mindset of using torchax.
Also, for existing users who already be required to activate the environment themselves, the extra activation introduces an overhead of state manipulation in PyTorch.

For example:

```python
# in class JittableModule
class JittableModule(torch.nn.Module):
  def functional_call(self, method_or_name, params, buffers, rng, *args, **kwargs):
    ...
      args = (self._model,) + args
    with (
      self._env.override_property(prng=rng),
      torch_stateless._reparametrize_module(self._model, params_copy),
    ):
      res = method(*args, **kwargs)
    return res

# in test_forward_reads_rng_from_env
class JittableModuleTest(unittest.TestCase):
  def test_forward_reads_rng_from_env(self):
    ...
    env = torchax.default_env()
    with env:
      jittable_module = interop.JittableModule(PlusOne().to("jax"))
      output = jittable_module(torch.ones(2, 2).to("jax"))

# This works even when we remove torchax.enable_globally()
```

In short, I'm worried about

* Unclear responsibility of activating the environment.
* Unnecessary overhead for existing users.
